apps/microsoft-teams/frontend/src/types/appActionResults.ts

```diff
@@ -0,0 +1,74 @@
+// Types that match the backend app action result schemas
+
+export interface ListChannelsResult {
+  ok: boolean;
+  data?: ChannelResult[];
+  error?: {
+    type: string;
+    message: string;
+    details?: Record<string, unknown>;
+  };
+}
+
+export interface ChannelResult {
+  id: string;
+  tenantId: string;
+  name: string;
+  teamId: string;
+  teamName: string;
+}
+
+export interface SendTestMessageResult {
+  ok: boolean;
+  data?: {
+    messageResponseId: string;
+  };
+  error?: {
+    type: string;
+    message: string;
+    details?: Record<string, unknown>;
+  };
+}
+
+// Type guards for runtime validation
+export function isListChannelsResult(value: unknown): value is ListChannelsResult {
+  if (typeof value !== 'object' || value === null) {
+    return false;
+  }
+
+  const obj = value as Record<string, unknown>;
+
+  return (
+    typeof obj.ok === 'boolean' &&
+    ((obj.data === undefined && obj.error !== undefined) ||
+      (obj.data !== undefined && obj.error === undefined))
+  );
+}
+
+export function isChannelResult(value: unknown): value is ChannelResult {
```

Comment:
Shower thought: As the app developer here I'm seeing a missed opportunity to align my front end and back end 🤔 .

It might not be realistic but it would be interesting to imagine:

* Declare my result schemas as constants somewhere. Share the code between app action and front end.
* For app action, use that result schema for the app action result schema that's bundled and pushed
* For front end, use typeschema and pass the same result schema to it to validate instead of handrolling these predicate functions.